Symbolic Paths

rgrinberg opened the pull request:
This is my attempt at porting to symbolic paths. Putting this up while this is WIP as I'm not sure I'm going in the right direction as I think you imagined this change being simpler than it's turning out for me here.

Repeating the question from slack: I'm not sure if Path.parent should escape the build dir for paths inside the build dir.

diml commented:
That seems fine to me, I expected the implementation of `Path` to change quite a bit. I think most of the complexity comes from the fact that paths inside the build directory are no longer always local and this needs to be handled correctly.

BTW, looking at how `Path.kind` and `Path.is_local` are used outside of `Path`, it seems that we often have two cases:
- either a path is an external path
- either a path is in the source tree or build directory

Before, the second class corresponded exactly to local paths, but this is no longer the case now. I think we should call this class *managed paths* and update the code accordingly, for instance with a  `Path.is_managed` function.

diml commented:
> Repeating the question from slack: I'm not sure if Path.parent should escape the build dir for paths inside the build dir.

It shouldn't. If we rely on `Path.parent Path.build_dir` being `Path.root` in some part of the code, we need to update the code to stop at `Path.build_dir`.

Note that this can potentially break user jbuild files if they try to escape the build directory to access the source tree directly.

rgrinberg commented:
> Before, the second class corresponded exactly to local paths, but this is no longer the case now. I think we should call this class managed paths and update the code accordingly, for instance with a Path.is_managed function.

I trid this change out locally and I think that it effectively obsoletes `Path.is_local`. I don't think there are any instances that rely on the path actually being external. Though I think there might be some cases where we can be a little more strict and explicitly check for `in_source_dir` vs. `in_build_dir`.

rgrinberg commented:
> Note that this can potentially break user jbuild files if they try to escape the build directory to access the source tree directly.

Yeah, but I think that's fine. I guess we should really be preventing the user from doing this by making sure the actions don't act on source dirs.

rgrinberg commented:
> It shouldn't. If we rely on Path.parent Path.build_dir being Path.root in some part of the code, we need to update the code to stop at Path.build_dir.

OK. How about we change `Path.parent` to be `t -> t option` to have a value for when we can't advance past root (with a `Path.parent_exn` for situations where we know this is allowed, perhaps).

diml commented:
> OK. How about we change Path.parent to be t -> t option to have a value for when we can't advance past root (with a Path.parent_exn for situations where we know this is allowed, perhaps).

That seems fine

diml commented:
Yh, I think we need to make this change

rgrinberg commented:
Ok, @diml this is getting close. Now the build dir is allowed to be external vs. local so it should be easy to just set it to the other thing. There are still some issues with other functions (see path.mlt), but I'm planning to fix those.

rgrinberg commented on src/stdune/path.ml:
```diff
@@ -221,86 +219,190 @@ module Local = struct
+    if String.is_prefix s ~prefix:"_build" then (
+      Exn.code_error "in_source_tree: path is in build dir"
+        [ "s", Sexp.To_sexp.string s ]
+    );
```
I'll get rid of these checks, they were just useful for debugging.

rgrinberg commented:
@diml latest is issue is with regards to configurator:

Configurator also uses path functions now, so it should set the build dir before doing so. I'm thinking that we should simply pass the build dir via `DUNE_CONFIGURATOR`. But to do that, we should really have a way of packing multiple values into that env var via sexps. Can we get the Sexp parsing stuff moved to Usexp yet?

rgrinberg commented:
Okay so `--build-dir` seems to finally work so reviewing can start. Sorry for the mess, I'll tidy up things whenever they don't make sense.

The failure on 4.02.3 is reproducible, but I don't yet understand what's causing it and why bootstrap works on other versions. Any ideas here would be welcome.

diml commented:
Maybe it's due to some change in the Unix library?

rgrinberg commented:
> Maybe it's due to some change in the Unix library?

It was just a bug on my part. The only reason it was only reproducible on 4.02.3 was because that's where we used sandboxes.

diml commented on src/stdune/path.ml:
```diff
@@ -221,86 +230,189 @@ module Local = struct
+let drop_build_dir p =
+  match build_dir_kind (), Kind.of_string p with
+  | Kind.External bd, Kind.External p ->
+    let open Option.O in
```
This seems a bit fragile to me. Because we don't normalize external paths, we might miss cases here. Why do we need this exactly?

rgrinberg replied:
If the user does something like `$ jbuilder build --build-dir /foo/bar /foo/bar/baz` I'd expect the path to be resolved to a build dir.

I also think that the db file also relies on this. Though we can fix that by changing the sexp reading/writing not to erase the the constructor of `Path.t`.

diml replied:
Yes, but what if the the user does `jbuilder build --build-dir /foo//bar /foo/bar/baz`?

diml commented:
The last two uses of `Path.kind` in the code base looks incorrect to me. In the two cases, we now want to match on wether the path is managed or not rather than local or external.

diml commented:
I think it would be easier to reason about the code if we have the guarantee that exactly one of the following assertion holds:

- either the build directory is an absolute path that doesn't point inside the source tree
- either it is relative to the source tree root

If the user passes an absolute path for the build directory, we don't check whether it points to the source tree or not.

I suggest to do the following: we implement a function similar to `realpath` to normalize absolute paths.  If the build directory is given as an absolute path, then we pass it through realpath, we pass the workspace root through realpath as well and we compare them. If the build directory is a descendant of the root, then we convert it to a local path.

rgrinberg commented:
> The last two uses of Path.kind in the code base looks incorrect to me. In the two cases, we now want to match on wether the path is managed or not rather than local or external.

Good catch, I'll fix those.

> If the user passes an absolute path for the build directory, we don't check whether it points to the source tree or not.

Shouldn't we verify this in set_build_dir? Otherwise we break your first bullet point: The build directory is an absolute path and it exists in the source directory.

diml commented:
> Shouldn't we verify this in set_build_dir? Otherwise we break your first bullet point: The build directory is an absolute path and it exists in the source directory.

I would think the right place is `Kind.of_string`.

rgrinberg commented on src/dune_project.ml:
```diff
@@ -105,7 +105,7 @@ end = struct
              |> String.concat ~sep:"/")
         in
-        if not (Path.is_local p) then invalid s;
+        if not (Path.is_managed p) then invalid s;
```
@diml I think this check and the one above can perhaps be tightened a little. E.g. `Path.is_in_source_tree` instead.

diml replied:
Indeed

rgrinberg commented:
@diml I've added a function that does normalization on external paths. It basically re-uses what you did for Local.relative. I'm not yet handling symlinks because some of the required code is in #769

rgrinberg commented:
@diml now external paths that point into the source dir should be resolved to local paths. There's still cleanup left to do, plus symlink handling.

rgrinberg commented:
@diml there's failures on windows because I'm not being careful enough with treating `strings` and `Local.t` as equivalent when they're really not. I will make `Local.t` abstract so that this error isn't possible anymore.

diml commented:
Seems good, and maybe `External.t` as well?

diml commented on src/stdune/path.ml:
```diff
@@ -2,17 +2,39 @@ let explode_path =
-    let acc = base :: acc in
     if dir = Filename.current_dir_name then
+      base :: acc
+    else if dir = path then
```
I remember doing similar things in the past but I don't remember why. What case does this catch?

rgrinberg replied:
It will not work for absolute paths otherwise. As `Filename.dirname "/" = "/"` it will just loop forever.

diml replied:
ok

rgrinberg commented:
> Seems good, and maybe External.t as well?

Will do but will try this in a separate PR if possible. This one has already accumulated so much stuff.

rgrinberg commented:
I've made External.t abstract as well. It was a bit of an eyesore when Local was abstract but External wasn't.

diml commented:
I'm just running benchmarks as part of the review since this is touching path.ml which is heavily used. I'm doing the following:

- I created a big workspace with all Jane Street packages + their dependencies
- I run `jbuilder build @install`
- I time a null build: `time jbuilder build -j1 @install`

I only compare the last step since a null build runs nothing and only setup the rules, so we effectively get the real overhead of Dune.

I get the following results on my machine:

| rev       | time    |
|-----------|---------|
| beta20 | ~2.2s |
| master | ~2.4s |
| this PR | ~2.9s |

This might be due to the extra allocations.

rgrinberg commented:
There's some low hanging fruit out here to cut the allocations down. For example, `Local.relative` is not really ideal.

diml commented:
I just realized that `Local.descendant` is wrong, the first then should return `root`...

rgrinberg commented:
I've pushed a few cleanups that could perhaps also help with the allocations. I see a couple of other sources of overhead that I've introduced:

* The round trip through to `Path.kind` for many basic operations
* Attempting to discover local paths as build paths.

Perhaps speeding those up could help a little.

diml commented:
Ok, I did the latter with by adding `Local.Prefix`

diml commented:
I tried `perf` but I can't see anything obvious. We could try another encoding that would avoid the indirection. For instance we define `Path.t` as `string` again and make path that are in the build directory start with byte '\x00'.

diml commented:
Regarding the slow-down in master, I did a bisection and it's caused by #666, which is not really surprising. I don't think we can do much about it now.

diml commented:
A tiny bit but not that much. We have a lot of structures indexed by paths, and path comparison is now slower. I suspect that this is the issue.

rgrinberg commented:
Hmm, so interning Local paths isn't as straightforward as I thought it would be. There's a couple of issues:

* Sorting on interned paths doesn't yield the results we'd expect. We really want to sort alphabetically here for the output.

* Incremental build are broken for some reason. E.g.:

```
$ jbuilder build # clean build
$ jbuilder build # full rebuild again
$ jbuilder build # now, nothing is rebuilt as expected.
```

rgrinberg commented:
@diml ok so incremental builds are now working again. I'm not really sure that the way I'm solving the problem with sorting is the right approach however.

 There's now 2 more issues:

* The output of dune changed needlessly. The order of the extensions is essentially random now. I think I can restore the old order if you'd like.

* There's 1 more failure which I don't yet understand:

```
./_build/default/bin/main.exe runtest --dev
     patdiff (internal) (exit 1)
(cd _build/default && /Users/rgrinberg/.opam/4.06.1/bin/patdiff -keep-whitespace -location-style omake -unrefined test/blackbox-tests/test-cases/misc/run.t test/blackbox-tests/test-cases/misc/run.t.corrected)
------ test/blackbox-tests/test-cases/misc/run.t
++++++ test/blackbox-tests/test-cases/misc/run.t.corrected
File "test/blackbox-tests/test-cases/misc/run.t", line 8, characters 0-1:
 |  $ dune runtest --display short
 |  File "dune", line 65, characters 21-44:
 |  Warning: Directory dir-that-doesnt-exist doesn't exist.
 |  No rule found for jbuild
 |          diff alias runtest
 |          diff alias runtest (exit 1)
 |  (cd _build/default && /usr/bin/diff -u result2 expected2)
-|  --- result2 2018-05-23 23:14:28.000000000 +0700
-|  +++ expected2       2018-05-23 23:14:28.000000000 +0700
+|  --- result2 2018-05-23 23:19:53.000000000 +0700
+|  +++ expected2       2018-05-23 23:19:53.000000000 +0700
 |  @@ -1 +1 @@
 |  -sub-tree/dir/b sub-tree/a
 |  \ No newline at end of file
 |  +sub-tree/a sub-tree/dir/b
 |  \ No newline at end of file
 |  [1]
```

Will investigate this tomorrow unless you have an idea what could be the issue.

rgrinberg commented:
Ah ok, the failure above is clear to me. We're iterating over an expanded glob and the order isn't deterministic. I'm not really sure what's the best way to fix this. I suppose we can fix the order, but I'm not really confident we're going to fix these problems everywhere. Another approach would be to just change `Pset.{iter,to_list,fold}` to give back the old order. This will result in a slow down and I'm still not 100% sure it will fix the issue everywhere, but I'm more confident about this.

rgrinberg commented:
@diml okay, so I've restored the old order for paths passed to the user. I'm not sure I did this exhaustively, but the tests seem to pass. It's probably a bad idea to introduce non-determinism to user builds so this needed fixing.

I also removed all the various aliases of Path.{Set,Map} if you don't mind. These identifiers aren't really long and grepping for them with all the aliases is really hard.

rgrinberg commented:
@diml looks like interning doesn't help the performance at all on my system. If anything, it makes it a bit slower. Quite a bit of overhead is spent in Interned.resize, which could be fixed by picking a more sensible initial table size for the path table.

rgrinberg commented:
@diml good news: the greedy resizing policy removes Table.resize from the profile. There's still more work to do to catch up with 1.20 however.

diml commented on src/stdune/path.ml:
```diff
@@ -490,29 +776,39 @@ let unlink_no_err t = try unlink t with _ -> ()
+    with
+    | Unix.Unix_error (EEXIST, _, _) -> ()
+    | Unix.Unix_error (ENOENT, _, _) ->
+      Exn.fatalf "Cannot create external build directory %s. \
```
We have `External.mkdir_p` now, should we use it here and in `Path.mkdir_p`?

rgrinberg replied:
It's not quite the same b/c mkdir_p would re-create the parents. What should this do when foo doesn't exist for example:

```
$ jbuilder build --build-dir /home/ocaml/foo/bar
```

diml replied:
I would want `foo` to be created.

diml commented on src/stdune/path.ml:
```diff
@@ -172,174 +268,353 @@ module Local = struct
+  (build_dir, build_dir_prefix, set_build_dir)
+
+module T : sig
+  type t = private
```
Why the `private`?

rgrinberg commented:
I use the constructor functions to help with debugging. I think I removed
those extra checks since, so we can get rid of it now.

On Tue, May 29, 2018, 12:40 AM Jérémie Dimino <notifications@github.com>
wrote:

> *@diml* commented on this pull request.
> ------------------------------
>
> In src/stdune/path.ml
> <https://github.com/ocaml/dune/pull/744#discussion_r191249038>:
>
> > @@ -490,29 +776,39 @@ let unlink_no_err t = try unlink t with _ -> ()
>
>  let build_dir_exists () = is_directory build_dir
>
> -let ensure_build_dir_exists () = Local.mkdir_p build_dir
> -
> -let extend_basename t ~suffix = t ^ suffix
> +let ensure_build_dir_exists () =
> +  match kind build_dir with
> +  | Local p -> Local.mkdir_p p
> +  | External p ->
> +    let p = External.to_string p in
> +    try
> +      Unix.mkdir p 0o777
> +    with
> +    | Unix.Unix_error (EEXIST, _, _) -> ()
> +    | Unix.Unix_error (ENOENT, _, _) ->
> +      Exn.fatalf "Cannot create external build directory %s. \
>
> We have External.mkdir_p now, should we use it here and in Path.mkdir_p?
> ------------------------------
>
> In src/stdune/path.ml
> <https://github.com/ocaml/dune/pull/744#discussion_r191250543>:
>
> > +  let build_dir = lazy (
> +    match !build_dir with
> +    | None ->
> +      Exn.code_error "build_dir: cannot use build dir before it's set" []
> +    | Some build_dir -> build_dir)
> +  in
> +  let build_dir_prefix = lazy (
> +    match !build_dir_prefix with
> +    | None ->
> +      Exn.code_error "build_dir: cannot use build dir before it's set" []
> +    | Some prefix -> prefix)
> +  in
> +  (build_dir, build_dir_prefix, set_build_dir)
> +
> +module T : sig
> +  type t = private
>
> Why the private?
>
> —
> You are receiving this because you authored the thread.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/dune/pull/744#pullrequestreview-123784671>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-1VrY2MJ-SUzG0goa0Zlh49AoIcKks5t3DZzgaJpZM4TyZwc>
> .
>

diml commented on bin/main.ml:
```diff
@@ -224,7 +227,9 @@ let common =
         display
+        build_dir
     =
+    let build_dir = Option.value ~default:"_build" build_dir in
```
Why not put the default value in the `Arg.(value ...)`?

rgrinberg replied:
I think it's because the default would have to be duplciated for bootstrap, but I'm no longer sure. I agree that making the default part of the CLI is correct.

rgrinberg commented on test/unit-tests/path.mlt:
```diff
@@ -312,3 +312,14 @@ Path.(reach_for_running (relative root "foo") ~from:(Path.relative root "foo"))
+|}]
+
+(* This is not right, but kind of annoying to fix :/ *)
+Path.relative (r "foo") "../_build"
```
Previously, I tried addressing this by having logic in the `in_source_tree` smart constructor to automatically change such paths to `In_build_dir`. This wasn't really necessary to make the feature work, so I dropped it. In any case, I wasn't comfortable with that approach either.

rgrinberg commented:
@diml I've read over your changes and I welcome them all. Github doesn't let me approve since I created this PR. I approve "manually".

diml commented:
Ok, I pushed a couple of additional changes. I let you review them

rgrinberg commented:
Okay, looks fine.

rgrinberg commented:
This is in master.
